src/main/java/org/codehaus/groovy/runtime/GStringImpl.java

```diff
@@ -47,18 +54,67 @@
     public GStringImpl(Object[] values, String[] strings) {
         super(values);
         this.strings = strings;
+        cacheable = checkValuesImmutable();
     }
 
     /**
      * Get the strings of this GString.
      * <p>
-     * This methods returns the same array as used in the constructor. Changing
-     * the values will result in changes of the GString. It is not recommended
-     * to do so.
+     * This methods returns the same array as used in the constructor.
+     * Changing the values will result in changes of the GString.
+     * It is generally not recommended to do so.
      */
     @Override
     public String[] getStrings() {
+        cacheable = false;
+        cachedStringLiteral = null;
         return strings;
     }
 
+    @Override
+    public Object[] getValues() {
+        cacheable = false;
+        cachedStringLiteral = null;
+        return super.getValues();
+    }
+
+    /**
+     * Get the strings of this GString as an unmodifiable list.
+     * This method is preferred over {@code getStrings()} in performance critical scenarios.
+     */
+    public List<String> getStringsList() {
+        return Collections.unmodifiableList(Arrays.asList(strings));
```

Comment:
How about simplifying `Collections.unmodifiableList(Arrays.asList(strings))` to `Arrays.asList(strings)` and renaming the method name to `getStringList()`?

Reply:
Forget it... I find the result of `Arrays.asList(strings)` supports changing the elements by `set` method.